driver-embedded/src/main/com/mongodb/embedded/client/EmbeddedInternalConnectionPool.java

```diff
@@ -0,0 +1,159 @@
+import com.mongodb.internal.connection.ConcurrentPool;
+import com.mongodb.internal.connection.InternalConnection;
+import com.mongodb.internal.connection.ResponseBuffers;
+import com.mongodb.session.SessionContext;
+import org.bson.ByteBuf;
+import org.bson.codecs.Decoder;
+
+import java.util.List;
+
+import static com.mongodb.assertions.Assertions.isTrue;
+
+class EmbeddedInternalConnectionPool {
+    private final ConcurrentPool<EmbeddedInternalConnection> pool;
+    private volatile boolean closed;
+
+    EmbeddedInternalConnectionPool(final EmbeddedInternalConnectionFactory internalConnectionFactory) {
+        this.pool = new ConcurrentPool<EmbeddedInternalConnection>(Integer.MAX_VALUE,
+                new EmbeddedConnectionItemFactory(internalConnectionFactory));
+    }
+
+    InternalConnection get() {
+        isTrue("Embedded connection pool is open", !closed);
+        return getPooledConnection();
+    }
+
+    void close() {
+        closed = true;
+        pool.prune();
+    }
+
+    private final class EmbeddedConnectionItemFactory implements ConcurrentPool.ItemFactory<EmbeddedInternalConnection> {
+        private final EmbeddedInternalConnectionFactory internalConnectionFactory;
+
+        EmbeddedConnectionItemFactory(final EmbeddedInternalConnectionFactory internalConnectionFactory) {
+            this.internalConnectionFactory = internalConnectionFactory;
+        }
+
+        @Override
+        public EmbeddedInternalConnection create(final boolean initialize) {
+            return internalConnectionFactory.create();
+        }
+
+        @Override
+        public void close(final EmbeddedInternalConnection embeddedInternalConnection) {
+            embeddedInternalConnection.close();
+        }
+
+        @Override
+        public ConcurrentPool.Prune shouldPrune(final EmbeddedInternalConnection embeddedInternalConnection) {
+            return closed ? ConcurrentPool.Prune.YES : ConcurrentPool.Prune.NO;
+        }
+    }
+
+    private InternalConnection getPooledConnection() {
+        PooledConnection connection = new PooledConnection(pool.get());
+        if (!connection.opened()) {
+            connection.open();
+        }
+        return connection;
+    }
+
+    private class PooledConnection implements InternalConnection {
+        private final EmbeddedInternalConnection wrapped;
+
+        PooledConnection(final EmbeddedInternalConnection wrapped) {
+            this.wrapped = wrapped;
+        }
+
+        @Override
+        public ConnectionDescription getDescription() {
+            return wrapped.getDescription();
+        }
+
+        @Override
+        public void open() {
+            wrapped.open();
+        }
+
+        @Override
+        public void openAsync(final SingleResultCallback<Void> callback) {
+            wrapped.openAsync(callback);
+        }
+
+        @Override
+        public void close() {
+            if (!wrapped.isClosed()) {
+                pool.release(wrapped);
```

Comment:
Allows the connection to be reused and the EmbeddedConnectionItemFactory is responsible for the actual closing of the EmbeddedInternalConnection.

Reply:
Fixed